### PR TITLE
Adding bundle version 4.0.2 to v3 and v4 images

### PR DESCRIPTION
**Pull request by soninaren**
## Screenshot of file structure

image

## Image size changes

| Image                                      |  Image ID       | Size     |
| -------------------------------------------| --------------- | -------- |
| 4-bullseye-python39.Dockerfile             |   3177160c0ed5  |  1.05 GB |
| 4-bullseye-python39-slim.Dockerfile        |   20f85cd1573b  |  1.04 GB |
| 4-bullseye-python38.Dockerfile             |   00587ffeb1f4  |  1.05 GB |
| 4-bullseye-python38-slim.Dockerfile        |   0446b24e10b8  |  1.04 GB |
| 4-bullseye-python37.Dockerfile             |   70aa46342077  |  1.05 GB |
| 4-bullseye-python37-slim.Dockerfile        |   0dbc785f5cbe  |  1.04 GB |
| 4-bullseye-python310.Dockerfile            |   d780366ef615  |  1.06 GB |
| 4-bullseye-python310-slim.Dockerfile       |   deb4dbd331f8  |  1.05 GB |
| 4-bullseye-powershell72.Dockerfile         |   e0b853075e4a  |  1.07 GB |
| 4-bullseye-powershell72-slim.Dockerfile    |   df5ff2411371  |  895 MB  |
| 4-bullseye-powershell70-slim.Dockerfile    |   51515824fc98  |  1.16 GB |
| 4-bullseye-powershell70.Dockerfile         |   51515824fc98  |  1.16 GB |
| 4-bullseye-node16-slim.Dockerfile          |   edbb3df6a632  |  859 MB  |
| 4-bullseye-node16.Dockerfile               |   edbb3df6a632  |  859 MB  |
| 4-bullseye-node14-slim.Dockerfile          |   ba2ffc5a8296  |  859 MB  |
| 4-bullseye-node14.Dockerfile               |   ba2ffc5a8296  |  859 MB  |
| 4-bullseye-java8-slim.Dockerfile           |   125b97ec8922  |  994 MB  |
| 4-bullseye-java8.Dockerfile                |   125b97ec8922  |  994 MB  |
| 4-bullseye-java11-slim.Dockerfile          |   7dd8b1473c67  |  1.11 GB |
| 4-bullseye-java11.Dockerfile               |   7dd8b1473c67  |  1.11 GB |
| 4-bullseye-dotnet-isolated-slim.Dockerfile |   32b487dce991  |  884 MB  |
| 4-bullseye-dotnet-isolated.Dockerfile      |   32b487dce991  |  884 MB  |
| 4-bullseye-dotnet-slim.Dockerfile          |   172a578d02d1  |  723 MB  |
| 4-bullseye-dotnet.Dockerfile               |   172a578d02d1  |  723 MB  |
| 3.0-buster-python39.Dockerfile             |   96d7429b0259  |  1.42 GB |
| 3.0-buster-python39-slim.Dockerfile        |   aba41870ef2a  |  1.08 GB |
| 3.0-buster-python38.Dockerfile             |   7a5280dcd9f7  |  1.42 GB |
| 3.0-buster-python38-slim.Dockerfile        |   50d814d85566  |  1.08 GB |
| 3.0-buster-python37.Dockerfile             |   bf902838bac1  |  1.42 GB |
| 3.0-buster-python37-slim.Dockerfile        |   a340eacd650d  |  1.08 GB |
| 3.0-buster-python36.Dockerfile             |   6a4a54f57b1a  |  1.4 GB |
| 3.0-buster-python36-slim.Dockerfile        |   9ac0ef2323a7  |  1.06 GB |
| 3.0-buster-powershell7-slim.Dockerfile     |   62bb33c9e1c6  |  1.19 GB |
| 3.0-buster-powershell7.Dockerfile          |   62bb33c9e1c6  |  1.19 GB |
| 3.0-buster-powershell6-slim.Dockerfile     |   a821968d0688  |  1.57 GB |
| 3.0-buster-powershell6.Dockerfile          |   a821968d0688  |  1.57 GB |
| 3.0-buster-node14.Dockerfile               |   f593b24bc5ff  |  1.46 GB |
| 3.0-buster-node14-slim.Dockerfile          |   ffb943d73b9f  |  1.17 GB |
| 3.0-buster-node12.Dockerfile               |   9559e53a9345  |  1.43 GB |
| 3.0-buster-node12-slim.Dockerfile          |   e24553bcd932  |  1.14 GB |
| 3.0-buster-node10.Dockerfile               |   d39b6b2f96a9  |  1.44 GB |
| 3.0-buster-node10-slim.Dockerfile          |   11a4c4f9e0b6  |  1.16 GB |
| 3.0-buster-java8-slim.Dockerfile           |   957173281ba6  |  1.2 GB |
| 3.0-buster-java8.Dockerfile                |   957173281ba6  |  1.2 GB |
| 3.0-buster-java11-slim.Dockerfile          |   376e606ec419  |  1.32 GB |
| 3.0-buster-java11.Dockerfile               |   376e606ec419  |  1.32 GB |
| 3.0-buster-dotnet-isolated-slim.Dockerfile |   269f2f1175d3  |  1.09 GB |
| 3.0-buster-dotnet-isolated.Dockerfile      |   269f2f1175d3  |  1.09 GB |
| 3.0-buster-dotnet-slim.Dockerfile          |   77a657063a5e  |  916 MB  |
| 3.0-buster-dotnet.Dockerfile               |   77a657063a5e  |  916 MB  |
| 3.0-bionic-dotnet.Dockerfile               |   f10d703570c4  |  917 MB  |

Releasing new Host versions.
[V3 Release](PASTE_V3_LINK_HERE)
[V2 Release](PASTE_V2_LINK_HERE)

---

### PR information

- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and CI is passing.

### CODE REVIEW

**Comment by pragnagopa** [comment]
@soninaren - please update PR description with changes in the Image size after including bundles V4

File: host/4/bullseye/amd64/dotnet/dotnet-inproc/dotnet-slim.Dockerfile [hunk 1]
```diff
@@ -28,6 +28,13 @@ RUN apt-get update && \
     unzip /$EXTENSION_BUNDLE_FILENAME_V3 -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V3 && \
     rm -f /$EXTENSION_BUNDLE_FILENAME_V3 &&\
     find /FuncExtensionBundles/ -type f -exec chmod 644 {} \;
+    EXTENSION_BUNDLE_VERSION_V4=4.0.2 && \
+    EXTENSION_BUNDLE_FILENAME_V4=Microsoft.Azure.Functions.ExtensionBundle.${EXTENSION_BUNDLE_VERSION_V4}_linux-x64.zip && \
+    wget https://functionscdn.azureedge.net/public/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V4/$EXTENSION_BUNDLE_FILENAME_V4 && \
+    mkdir -p /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V4 && \
+    unzip /$EXTENSION_BUNDLE_FILENAME_V4 -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V4 && \
```
Comment (pragnagopa) [inline host/4/bullseye/amd64/dotnet/dotnet-inproc/dotnet-slim.Dockerfile:36]: Please run docker image locally and verify that this only copies AMD64 bits

**Comment by pragnagopa** [comment]
Adding notes from @mattchenderson from offline conversation:

```
But just as a curiosity, I decided to do a quick analysis from within a container:
du --summarize azure-functions-host FuncExtensionBundles -h
260M    azure-functions-host
401M    FuncExtensionBundles

So I think the host has some opportunity, but also the Bundles, which include all the runtime targets. I assume this is because we’re not doing split manifests by platform, and maybe we should.
86M     FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/3.15.0/bin/runtimes
86M     FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/2.17.0/bin/runtimes

ls FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/3.15.0/bin/runtimes
linux-arm64  linux-x64  osx-x64  unix  win  win-arm  win-arm64  win-x64  win-x86  win7-x64

And within those, it looks like there’s a lot of chunky native dependencies (namely SQL clients) brought in…

But there’s also… what looks to be V3 stuff baked in as well?
83M     FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/2.17.0/bin_v3
91M     FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/3.15.0/bin_v3

If I rip out V3 and everything that isn’t, let’s say… linux-x64, then we’re talking about a savings of ~301MB. That’s a start, maybe.

```

**Review by pragnagopa — CHANGES_REQUESTED** [review]
@soninaren - Please 
- Capture Image size changes
- Only add AMD64 binaries
- Delete other runtimes from V2 and V3 bundles

**Comment by soninaren** [comment]
@pragnagopa, have made the following changes, and captured the information as requested

* Delete other runtimes from V2 and V3 bundles
* Only add AMD64 binaries
-> Removed bin directory as it was not needed for v3 and v4 images.

* Capture Image size changes

| Images | Description |Size |
| --- | --- | --- |
| 3.0\buster\amd64\dotnet\dotnet-inproc\dotnet.Dockerfile | Includes bundle v4 (with bin directory deleted) | 928 MB |
| 3.0\buster\amd64\dotnet\dotnet-inproc\dotnet.Dockerfile| Does not include bundle v4 | 1.15 GB |
| 4\bullseye\amd64\dotnet\dotnet-inproc\dotnet.Dockerfile| Includes bundle v4 (with bin directory deleted) | 735 MB |
| 4\bullseye\amd64\dotnet\dotnet-inproc\dotnet.Dockerfile| Does not include bundle v4 | 860 MB |

**Comment by pragnagopa** [comment]
@soninaren - thanks for the updates!

Please update the PR description to include sizes for all the images as this PR is about including new version across all of them.

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile [hunk 2]
```diff
@@ -22,18 +22,28 @@ RUN EXTENSION_BUNDLE_VERSION=1.8.1 && \
     mkdir -p /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION && \
     unzip /$EXTENSION_BUNDLE_FILENAME -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION && \
     rm -f /$EXTENSION_BUNDLE_FILENAME && \
+    rm -r /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/${EXTENSION_BUNDLE_VERSION}/bin && \
     EXTENSION_BUNDLE_VERSION_V2=2.19.0 && \
     EXTENSION_BUNDLE_FILENAME_V2=Microsoft.Azure.Functions.ExtensionBundle.${EXTENSION_BUNDLE_VERSION_V2}_linux-x64.zip && \
     wget https://functionscdn.azureedge.net/public/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V2/$EXTENSION_BUNDLE_FILENAME_V2 && \
     mkdir -p /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V2 && \
```
Comment (pragnagopa) [inline host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile:29]: from offline discussion - please see if we can delete V2 all together

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile [hunk 3]
```diff
@@ -22,18 +22,28 @@ RUN EXTENSION_BUNDLE_VERSION=1.8.1 && \
...
     wget https://functionscdn.azureedge.net/public/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V3/$EXTENSION_BUNDLE_FILENAME_V3 && \
     mkdir -p /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V3 && \
     unzip /$EXTENSION_BUNDLE_FILENAME_V3 -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V3 && \
     rm -f /$EXTENSION_BUNDLE_FILENAME_V3 &&\
+    rm -r /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/${EXTENSION_BUNDLE_VERSION_V3}/bin && \
+    EXTENSION_BUNDLE_VERSION_V4=4.0.2 && \
+    EXTENSION_BUNDLE_FILENAME_V4=Microsoft.Azure.Functions.ExtensionBundle.${EXTENSION_BUNDLE_VERSION_V4}_linux-x64.zip && \
+    wget https://functionscdn.azureedge.net/public/ExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V4/$EXTENSION_BUNDLE_FILENAME_V4 && \
+    mkdir -p /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V4 && \
+    unzip /$EXTENSION_BUNDLE_FILENAME_V4 -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION_V4 && \
```
Comment (pragnagopa) [inline host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile:44]: For completeness - please run docker file locally , capture screen shot of directories in bundles and paste in the PR description - to ensure we only have binaries needed and rest of them are cleaned up. Please capture this for one of the images - no need to collect this info across all of them

**Comment by pragnagopa** [comment]
@CooperLink - please review!

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile (hunk 2 above)
Comment (soninaren) [reply host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile:29]: This PR does delete netcore 2.0 binaries. The bin directory contains the v2 binaries. I have deleted it from all bundles

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile (hunk 3 above)
Comment (soninaren) [reply host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile:44]: Sure, I did verify locally that it is deleted. I can capture the screenshot as well

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile (hunk 2 above)
Comment (CooperLink) [reply host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile:29]: What are we getting from v1/v2 extension bundle commands here? It kinda looks to me like we are downloading them, unzipping them, then deleting them? I am sure that is not the case. But, what content does our final image get from the v1/v2 cmds here?

**Review by CooperLink — APPROVED** [review]

**Review by VpOfEngineering — APPROVED** [review]
LGTM

File: host/4/bullseye/amd64/dotnet/dotnet-inproc/dotnet-slim.Dockerfile (hunk 1 above)
Comment (soninaren) [reply host/4/bullseye/amd64/dotnet/dotnet-inproc/dotnet-slim.Dockerfile:36]: This is complete

File: host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile (hunk 2 above)
Comment (soninaren) [reply host/3.0/bionic/arm32v7/dotnet/dotnet.Dockerfile:29]: The unzipped content of the bundle zip files ends up in the final image. It contains binaries and template files that are referenced by the function runtime based on customer's host.json configuration.

**Comment by soninaren** [comment]
@pragnagopa can you approve?

**Review by pragnagopa — APPROVED** [review]
Please capture size for each image and also screen shot of the directories in extension bundles

**Comment by soninaren** [comment]
> Please capture size for each image and also screen shot of the directories in extension bundles

Thanks @pragnagopa

I have information from 27 out of 51 images. Rest of the images are still building, will add the information as soon as all images are built.

**Pull request merged**